### ISSUE TITLE
[HOLD for payment 2023-08-24] Show chats in the LHN that have at least 1 ADDComment action, or 1 draft message

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#14523]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Login with any account
3. Search for a user that you don't have any messages with, open the chat but don't send any message

## Expected Result:
Chat should not show in the LHN as there is not at least 1 ADDComment action, or 1 draft message.

## Actual Result:
Empty chat shows in the LHN and remains in the LHN after simply navigating to it.

## Workaround:
None

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.58-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**
**Issue reported by:**
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fbd874b8451c4608</li>
        <li>Upwork Job ID: 1618016837618089984</li>
        2023-01-24<li>Automatic offers: </li>
<ul>
<li> |  | 0</li>
</ul></li>
    </ul>
</details>

**Comment by JmillsExpensify** [comment]
Pretty sure this is internal based on the [this discussion](https://github.com/Expensify/App/issues/11633), so I'm starting with that route first for now.

**Comment by JmillsExpensify** [comment]
Huh, I'm just seeing that now one internal was added to this issue. I'm not sure why that's the case. Trying to more labels to hopefully get it right.

**Comment by JmillsExpensify** [comment]
@PauloGasparSv Do you mind weighing in on this issue and whether we can open it up for external? Thank you!

**Comment by PauloGasparSv** [comment]
Hey @JmillsExpensify I'm pretty sure this can be an External change. I think we can add a filter in [shouldReportBeInOptionList](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js#L1139) that filters out empty chats. That method is already called to filter out chats in LNH and the Search option (before you type anything) 

But I'm not sure which fields to check so I created a [thread on slack](https://expensify.slack.com/archives/C03TQ48KC/p1675266381428049) to discuss that

**Comment by JmillsExpensify** [comment]
Ok great! I'll add the `external` label in the meantime, looks like there is largely agreement in that thread.

**Comment by Pujan92** [comment]
**Proposal**

We can add last condition to check the `hasDraft` or `lastMessage` for the report type `Chat Report` and based on that we can filter out those entries within `shouldReportBeInOptionList`.

```diff
diff --git a/src/libs/ReportUtils.js b/src/libs/ReportUtils.js
index 4bda454798..01a2dc2387 100644
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -1194,6 +1194,10 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
 
+    if (_.isEqual(report.reportName, CONST.REPORT.DEFAULT_REPORT_NAME) && !report.hasDraft && (_.isEmpty(report.lastMessageTex
+ t) || (_.isEmpty(report.lastMessageHtml)))) {
+        return false;
+    }
+
     return true;
 }
```

**Comment by bernhardoj** [comment]
### **Proposal**

The first thing we need to do is to check whether the given report is a direct message or not. Next, we can simply check whether the `lastMessageText` of the report is empty or not. We already have the filter for draft report.

https://github.com/Expensify/App/blob/7f78d52b032b9c8abab2b0b5f10081a00f34b28b/src/libs/ReportUtils.js#L1229-L1233

Based on the condition above, empty pinned chat will still be shown.

```diff
diff --git a/src/libs/ReportUtils.js b/src/libs/ReportUtils.js
index 4bda45479..d0ff1c66a 100644
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -142,6 +142,10 @@ function canDeleteReportAction(reportAction) {
         && reportAction.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE;
 }
 
+function isDirectMessage(report) {
+    return _.isEmpty(getChatType(report));
+}
+
 /**
  * Whether the provided report is an Admin room
  * @param {Object} report
@@ -1194,6 +1198,10 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
 
+    if (isDirectMessage(report) && _.isEmpty(report.lastMessageText)) {
+        return false;
+    }
+
     return true;
 }
 
```

**Comment by Pujan92** [comment]
**Updated Proposal**

As draft message we are already checking in earlier condition and @bernhardoj also mentioned, we do not require to add it here.

```diff
diff --git a/src/libs/ReportUtils.js b/src/libs/ReportUtils.js
index 4bda454798..01a2dc2387 100644
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -1194,6 +1194,10 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
     
+    if (_.isEqual(report.reportName, CONST.REPORT.DEFAULT_REPORT_NAME) && _.isEmpty(report.lastMessageText)) {
+        return false;
+    }

```

**Comment by sobitneupane** [comment]
**Proposal:**
If we don't want the reports to be in LHN even when the report is opened, we need to remove  `if (report.reportID === reportIDFromRoute) {return true;}` condition.

```diff
@@ -1145,13 +1145,6 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
 
-    // Include the currently viewed report. If we excluded the currently viewed report, then there
-    // would be no way to highlight it in the options list and it would be confusing to users because they lose
-    // a sense of context.
-    if (report.reportID === reportIDFromRoute) {
-        return true;
-    }
-
     // Include reports if they have a draft, are pinned, or have an outstanding IOU
     // These are always relevant to the user no matter what view mode the user prefers
     if (report.hasDraft || report.isPinned || hasOutstandingIOU(report, currentUserLogin, iouReports)) {
@@ -1194,6 +1187,10 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
 
+    if (_.isEmpty(getChatType(report)) && _.isEmpty(report.lastMessageText)) {
+        return false;
+    }
+
     return true;
 }
```

**Comment by PauloGasparSv** [comment]
Let's wait on @JmillsExpensify to confirm that
IMO it will make the solution a bit more complicated and won't aggregate much value to the App!

Also, let me bring a bit of the slack discussion about which fields we should use to check if the chat is empty to this GH:

> If all you have is the report object itself then I’d say the best way to confirm if a chat is “empty” would be to look at the reportActionCount. Check where we return the reports to the client in [Auth here](https://github.com/Expensify/Auth/blob/f5d829f716a708305751fc42c49e0cc2131a2a3c/auth/command/GetChats.cpp#L91-L102). One caveat is that IIRC we are going to backfill all reports so that they have at least 1 CREATED action so an empty chat would be reportActionCount === 1 .

I think we should be checking the `reportActionCount` field instead of `lastMessageHtml` or `lastMessageText` cc @Pujan92 @bernhardoj

**Comment by PauloGasparSv** [comment]
Just confirmed with the team the `reportActionCount` should start at `1` now that we [ensure every report created inserts a CREATED action](https://github.com/Expensify/Expensify/issues/245771)!

So I think we should be skipping chats if `reportActionCount === 1` as mentioned above!

**Comment by Pujan92** [comment]
@PauloGasparSv I am not seeing the field `reportActionCount` while logging the report details which we have while filtering.

<img width="845" alt="Screenshot 2023-02-02 at 9 25 20 PM" src="https://user-images.githubusercontent.com/14358475/216374737-76ee19bf-6e54-4bbe-a809-11f20c0bcfbc.png">

**Comment by bernhardoj** [comment]
Yeah, looks like the `reportActionCount` is not present to the report data. Assuming it is a new property that is not deployed yet, we can simply change `_.isEmpty(report.lastMessageText)` to `report.reportActionCount === 1`.

**Comment by PauloGasparSv** [comment]
You are correct! Sorry for that @Pujan92 @bernhardoj 

Let me check why that property is not returned and if I can make it available.

**Comment by PauloGasparSv** [comment]
All right, just checked and while formatting the field `reportActionCount` is returned as `maxSequenceNumber`.

Could everyone please update the proposals to use that field instead to check if the chat is empty making sure the proposal still works and that counter starts at 1?

**Comment by Pujan92** [comment]
**Updated Proposal**

Use `maxSequenceNumber` number to check for the empty chat report.

```diff
diff --git a/src/libs/ReportUtils.js b/src/libs/ReportUtils.js
index 4bda454798..01a2dc2387 100644
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -1194,6 +1194,10 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
 
+    if (_.isEmpty(report.reportType) && report.maxSequenceNumber === 1) {
+        return false;
+    }
+
     return true;
 }
```

**Comment by sobitneupane** [comment]
@PauloGasparSv I don't think we can use `maxSequenceNumber`, if we add a comment and delete it, `maxSequenceNumber` number increases and is not equal to 1 though the report has no comments.

**Comment by PauloGasparSv** [comment]
In the meantime, created a [fix](https://github.com/Expensify/Web-Expensify/pull/36350) so the `maxSequenceNumber` is not coalesced to 0 but instead to 1 in case it is missing.

Bump @JmillsExpensify on [this](https://github.com/Expensify/App/issues/14523#issuecomment-1413896711).

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv I don't think we can use maxSequenceNumber, if we add a comment and delete it, maxSequenceNumber number increases and is not equal to 1 though the report has no comments.

Nice catch @sobitneupane, I'll bring it up on the field discussion thread that suggested it!
But I think it's ok to show the chat in that edge case, how about you @JmillsExpensify @thesahindia ?

**Comment by PauloGasparSv** [comment]
CC @marcaaron would you mind sharing your opinion on [this](https://github.com/Expensify/App/issues/14523#issuecomment-1414135592) too?

This issue is about not showing the chat in LNH if it has no reportActions or drafts in it. The field `reportActionCount` / `maxSequenceNumber` doesn’t decrease back to 1 after deleting everything in the chat, so it's not an absolute way to check if the chat is empty using just the report object.

 I think this would be an edge case and would be ok to have, do you?

**Comment by marcaaron** [comment]
>not showing the chat in LNH if it has no reportActions or drafts in it

yes, sounds like the expected behavior to me, but don't have any super strong feelings on it. I think I'd add that the "active" chat in view will still show whether there's a draft or not.

Also, all reports with have at least one `'CREATED'` action soon so not sure how reliable it will be to check for `reportActions.length === 0`. In addition, the `reportActions` are not all loaded by default. They are paginated so you can't really depend on the stored `reportActions` at all.

>The field reportActionCount / maxSequenceNumber doesn’t decrease back to 1 after deleting everything in the chat, so it's not an absolute way to check if the chat is empty using just the report object.

Cool. I forgot that this doesn't include deleted actions (I think that's the correct behavior).

>I think this would be an edge case and would be ok to have, do you?

What's the worst case scenario? Every edge case is better evaluated as "what will the user experience if they do run into it". Sounds like the chat would just show in the LHN? Agree. That it doesn't sound so bad.

But if we do want to fix it then we can return from Auth a `visibleActionCount` or number of visible and non-deleted actions and update it as things are deleted client side.

**Comment by bernhardoj** [comment]
###  **Updated Proposal**

Updating proposal based on the discussion above.

```diff
diff --git a/src/libs/ReportUtils.js b/src/libs/ReportUtils.js
index 4bda45479..d0ff1c66a 100644
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -142,6 +142,10 @@ function canDeleteReportAction(reportAction) {
         && reportAction.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE;
 }
 
+function isDirectMessage(report) {
+    return _.isEmpty(getChatType(report));
+}
+
 /**
  * Whether the provided report is an Admin room
  * @param {Object} report
@@ -1194,6 +1198,10 @@ function shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, curr
         return false;
     }
 
+    if (isDirectMessage(report) && report.maxSequenceNumber === 1) {
+        return false;
+    }
+
     return true;
 }
 ```

**Comment by PauloGasparSv** [comment]
Bump @JmillsExpensify @thesahindia

**Comment by JmillsExpensify** [comment]
> I think the chat should be present in LHN when it is open.

I don't think I agree with this, though in any case it's also not how it works today, so I think that's out of scope for this issue.

**Comment by bernhardoj** [comment]
Actually, that's how it works currently. To test it, you can open any chat first, and then switched to #focus mode then open another chat. Notice the chat from the LHN is getting removed after you moved to another chat.

https://github.com/Expensify/App/blob/7f78d52b032b9c8abab2b0b5f10081a00f34b28b/src/libs/ReportUtils.js#L1222-L1227

**Comment by MelvinBot** [comment]
@JmillsExpensify @PauloGasparSv @thesahindia this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by JmillsExpensify** [comment]
I replied to this question [two days ago here](https://github.com/Expensify/App/issues/14523#issuecomment-1419162097). To repeat that comment, I don't agree that a chat should be present in the LHN when it is open.

**Comment by JmillsExpensify** [comment]
Actually this is interesting. I just realized what a couple of days ago I said what I said. On desktop, we don't show the open chat in the LHN. If you're on web though, we do show the open chat in the LHN. So we're actually not consistent cross-platform at the moment.

**Comment by JmillsExpensify** [comment]
@thesahindia do you mind bringing this one to open-source tomorrow. Given that we don't have a standard at the moment, we should agree what the standard is.

**Comment by PauloGasparSv** [comment]
Not overdue, slack [discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1676043596765619) still ongoing

**Comment by JmillsExpensify** [comment]
Thanks for your patience everyone. We've aligned on [this behavior](https://expensify.slack.com/archives/C01GTK53T8Q/p1676302346511899?thread_ts=1676043596.765619&cid=C01GTK53T8Q), so we should be good to continue with proposals again.

**Comment by JmillsExpensify** [comment]
And...GH just updated after I commented. We'll wait for @PauloGasparSv to weigh in when he returns tomorrow. Thanks again for the patience everyone!

**Comment by MelvinBot** [comment]
📣 @bernhardoj You have been assigned to this job by @PauloGasparSv!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by PauloGasparSv** [comment]
Agree with @thesahindia and their reason, assigned @bernhardoj here

**Comment by bernhardoj** [comment]
Thanks! I will open the PR in a few hours.

**Comment by bernhardoj** [comment]
PR is ready for review.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.72-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15129


If no regressions arise, payment will be issued on 2023-02-23. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@thesahindia / @PauloGasparSv] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@thesahindia / @PauloGasparSv] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@thesahindia / @PauloGasparSv] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@bernhardoj] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@JmillsExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by JmillsExpensify** [comment]
Woo! That was fast. PR qualifies for the 50% bonus, so I believe this is where that leaves us:

 - Issue reporter: Expensify $0
 - Contributor: @bernhardoj $1500 (incl. 50% bonus)
 - Contributor+: @thesahindia $1500 (incl. 50% bonus). 

Sound right? The upwork job is here: https://www.upwork.com/jobs/~01fbd874b8451c4608. Please apply and we can finalize the final payment when the regression period ends. Thanks all!

**Comment by bernhardoj** [comment]
Applied.

Here is the Regression Test Steps:
1. Start a new chat with a user that you don't have any messages with
2. Switch to another chat
3. Verify the empty new chat disappear from LHN

Do we agree 👍 or 👎

**Comment by PauloGasparSv** [comment]
Still need to investigate but maybe the changes from here created https://github.com/Expensify/App/issues/15251

I did test that issue's scenario (sending a comment in the Chat to see if it would stick in the LNH), and it did work so I'm confused. I still didn't replicate it locally and a comment there suggests the problem is in the backend, planning on investigating it on Monday.

**Comment by JmillsExpensify** [comment]
@bernhardoj I think that addition looks good. I agree with you that everything else should already be covered for the other cases, including an existing chat and a draft message.

**Comment by roryabraham** [comment]
I just created https://github.com/Expensify/App/pull/15334 to revert the solution for this issue, because:

- it relied on sequence numbers which have been completely deprecated
- it caused [a regression](https://github.com/Expensify/App/issues/15251)

I think this is more fair than proceeding with https://github.com/Expensify/App/issues/15251, because we should not pay a separate bounty to resolve the regression. Instead we should revert the PR that introduced the regression and try again with a proper fix.

**Comment by roryabraham** [comment]
There is a front-end piece that @bernhardoj can still handle to earn the bounty here, but there's also some back-end stuff we'll need to do. I will start working on writing up a plan of action.

**Comment by JmillsExpensify** [comment]
Sounds like a plan Rory. I'll update the issue to clarity that we're HOLDing on a regression fix.

**Comment by roryabraham** [comment]
Okay, it's basically a mini design doc, but here's my proposed plan

---

# Proposal
1. Start maintaining an accurate `reportActionCount` field
1. Don't show DM's with only 1 reportAction (i.e: the `CREATED` action) in options list, unless it is the currently viewed report

# Please re-state the problem that we are trying to solve in this issue.
When you create a new DM but don't send any messages, it should not appear in the LHN for other users. If you don't create a draft message, then navigate away from the chat, it should disappear from the LHN for you.

# What is the root cause of that problem?
(after https://github.com/Expensify/App/pull/15334 is merged) There is no special logic in `ReportUtils.shouldReportBeInOptionsList` to hide DMs with only a created action from the LHN – the default is to show the report.

# What changes do you think we should make in order to solve the problem?

## Add a `reportActionCount` field to reports in Onyx

To make this change, we need to examine all NewDot flows that add or remove reportActions from a report. This includes:

- Creating a new report (adds a `CREATED` action)
- Sending a message or attachment
- Sending an IOU
- Deleting a reportAction

### Propagating `reportActionCount` correctly in the API

First, we need to make sure that when we initially fetch reports, those reports contain their canonical `reportActionCount` from the back-end. Fortunately, the `reportActionCount` is already returned from `Auth::GetChats`, `Auth::GetReportSummaryList`, `Auth::CreateChatReports`, etc..., so all we have to do for this is to update [`ReportUtils::structureReportForOnyx`](https://github.com/Expensify/Web-Expensify/blob/12106f5130cb4f0883675d1129a398f12b7dcdab/lib/ReportUtils.php#L2418) to include `reportActionCount`, which, as @PauloGasparSv pointed out above, is already used to generate the `maxSequenceNumber` field.

Next, we'll need to update `ReportUtils::buildOnyxDataForNewAction` [here](https://github.com/Expensify/Web-Expensify/blob/12106f5130cb4f0883675d1129a398f12b7dcdab/lib/ReportUtils.php#L1815-L1829) so that we include `reportActionCount` as well as the other fields.

Lastly, we'll update [`ReportAPI::editComment`](https://github.com/Expensify/Web-Expensify/blob/41873abd533de094c0e38387e6689b17c03e4548/lib/ReportAPI.php#L492-L623), so that, if we're deleting a comment we always update the report in Onyx to decrement the `reportActionCount`, regardless for whether it's the newest visible reportAction on the report, or if we `shouldKeyReportActionsByID`.

### Optimistically updating reports with the correct `reportActionCount`

#### Creating a new report
There are a few locations where new reports are created in the NewDot front-end, and we can find them by searching for usages of `ReportUtils.buildOptimisticChatReport`:

- In [`src/libs/actions/Report.navigateToAndOpenReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L411)
- In [`src/libs/actions/Report.addPolicyReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L959)
- In [`ReportUtils.buildOptimisticWorkspaceChats`](https://github.com/Expensify/App/blob/a5b9c9ba88fa0634a963bd248152b79e3006bd6d/src/libs/ReportUtils.js#L1134)
- In [`IOU.requestMoney`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/IOU.js#L63)
- In [`IOU.createSplitsAndOnyxData`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/IOU.js#L232)
- In [`IOU.getSendMoneyParams`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/IOU.js#L676)

In each of the above-listed locations, there is some code to add an optimistic `CREATED` action, along with potentially more actions such as an IOU action. We will want to trace each of these flows and ensure that we set the correct optimistic `reportActionCount` according to the number of optimistically-generated `reportActions` before writing the report to Onyx.

- The optimistic `reportActionCount` can safely be assumed to be `1` in each of these places:
    - [Here in `openReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L373-L395)
    - [here in `addPolicyReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L956-L1027)
    - For each optimistic report in [`buildOptimisticWorkspaceChats`](https://github.com/Expensify/App/blob/a5b9c9ba88fa0634a963bd248152b79e3006bd6d/src/libs/ReportUtils.js#L1134)
- In `IOU.requestMoney` and `IOU.getSendMoneyParams`, the optimistic `reportActionCount` should be _incremented_ by `1` if we are not creating a new report, but should be `2` if we are creating a new report
- `IOU.createSplitsAndOnyxData` should use the same defaults as `IOU.requestMoney`, except it will vary for the group DM and for every 1:1 DM associated with the bill split.

#### Sending a message / attachment
This happens in `Report.addActions`, where we'll always be able to assume that the report has an existing `reportActionCount` field. We will increment that field by `1` if there is just text or just an attachment, but `2` if there is both text and an attachment.

#### Sending an IOU
This is covered by the same flows listed above under `Creating a new report`. If creating a new report, the correct value for `reportActionCount` on the new report is `2`. If adding an `IOU` action to an existing report, the `reportActionCount` should be incremented by `1`.

#### Deleting a reportAction
This flow lives in [`src/libs/actions/Report.deleteReportComment`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L678-L751). The report is optimistically updated [here](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L696-L701), so we need to update that piece of code to also optimistically decrement the `reportActionCount` on the report by `1`.

## Don't show DM's with only 1 reportAction in options list (unless it is the currently viewed report)

To do this, we will draw inspiration from https://github.com/Expensify/App/pull/15129: 

1. Add a function called `isDirectMessage` to `src/libs/ReportUtils` that returns `true` if the `chatType` rNVP on the report is empty, and `false` if it is not.
1. Update `src/libs/ReportUtils.shouldReportBeInOptionList` to return `false` if the report is a direct message and the `reportActionCount` is `1`.

# Proposed rollout plan
1. Make the PHP changes to add the `reportActionCount` field into reports in Onyx. This can probably be 1 PR, but could also be split up. It should be covered by unit tests.
1. [HOLD 1] Optimistically add `reportActionCount` in `openReport`
1. [HOLD 1] Optimistically add `reportActionCount` in `addPolicyReport`
1. [HOLD 1] Optimsitically add `reportActionCount` in `buildOptimisticWorkspaceChats`
1. [HOLD 2] Optimistically add `reportActionCount` in `addActions`
1. [HOLD 5] Optimistically add `reportActionCount` in `deleteReportComment`
1. [HOLD 6] Optimistically add `reportActionCount` in `requestMoney`
1. [HOLD 7] Optimistically add `reportActionCount` in `getSendMoneyParams`
1. [HOLD 7] Optimistically add `reportActionCount` in `createSplitsAndOnyxData`
1. [HOLD 1-9] Don't show DM's with only 1 reportAction in options list (unless it is the currently viewed report)

---

👏🏼 So given that this is a broad issue with a large scope, I suggest we create a GitHub project or tracking issue to track these changes, with a separate issue for each item in the rollout plan above. This particular issue will basically be the issue for item 10 in the proposed rollout plan above.

Does this sound like a good plan @PauloGasparSv @JmillsExpensify @chiragsalian?

**Comment by bernhardoj** [comment]
Sorry for the regression guys.

I like the plan above, I can handle the frontend part. I would also like to add that because we will optimistically increment the `reportActionCount`, we should decrement it back if it fails.

For example, when we send a single message, we will increment it by 1, but when it fails for some reason, we decrement it by 1.

**Comment by MelvinBot** [comment]
📣 @bernhardoj! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by bernhardoj** [comment]
Contributor details
Your Expensify account email: bernhard.josephus@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~013e648527fa0266fd

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by chiragsalian** [comment]
I like the proposed plan rory 👍 

My one concern is the optimistic plan. In the rollout plan I'm wondering if before step 2 we should create a centralized method like `getReportActionCount` that connects to the report and returns the `reportActionCount`. This way all the other places can use this method and then increment or decrement the number as needed.

Let's remember for every optimistic action. We should also add failure data to revert the value as necessary.

I'm not sure i follow why some of the flows are on HOLD for steps other than 1. 

And i agree we should make another tracking issue.

**Comment by JmillsExpensify** [comment]
All good @bernhardoj! Thanks @roryabraham for the investigation and getting this mini-plan typed up.

**Comment by JmillsExpensify** [comment]
Ya'll want me to take on creating the tracking issue and related issues?

**Comment by PauloGasparSv** [comment]
The proposal is amazing, thanks so much for the help @roryabraham!


>  Ya'll want me to take on creating the tracking issue and related issues?

@JmillsExpensify that would be great! Would you mind creating those?

**Comment by roryabraham** [comment]
> I would also like to add that because we will optimistically increment the reportActionCount, we should decrement it back if it fails.

Correct @bernhardoj! When dealing with `API.write` commands, we have `optimisticData`, `successData` and `failureData`. What you've described is a good use-case for `failureData` 👍🏼

**Comment by JmillsExpensify** [comment]
Nice, I'll do that now ya'll.

**Comment by JmillsExpensify** [comment]
A couple of questions follow:

 - There's a lot of internal implementation, so who's going to take these on? Someone in this issue, or should we raise it in #engineering-chat in Slack? I went ahead and assigned myself for now. 
 - Shouldn't we post on this more broadly in open-source somewhere?

Oh and finally, perhaps we'll put this issue on hold now for the tracking issue?

Once we align on the above I can clean up some of the labels and what not on the implementation issues. Thanks!

**Comment by chiragsalian** [comment]
> Someone in this issue, or should we raise it in #engineering-chat in Slack?

Anyone can. if we go with the internal route i prefer if we post on slack so that it's spread out faster.

> Shouldn't we post on this more broadly in open-source somewhere?

I think all the App issues posted above can be handled externally. We just have to share steps including failure steps with our expected outcome. Maybe @roryabraham can confirm since I'm unsure why he suggested all internal earlier.

> Oh and finally, perhaps we'll put this issue on hold now for the tracking issue?

I'd say we can just close this since the last issue mentioned [here](https://github.com/Expensify/App/issues/15352) should tackle this issue. If it hasn't then we reopened this issue later.

Also @roryabraham, did you have a chance to check out my [comment](https://github.com/Expensify/App/issues/14523#issuecomment-1439329898) asking for
1. Clarification of the HOLD status for the ones not on HOLD for 1. I don't see why they are on HOLD since each could be worked on in parallel as there shouldn't be any impact on the UI except for the last issue. For example, for `deleteReportComment` we could add to the steps to fake insert reportActionCount 10 and then the expected optimistic data should decrease the number and its failure data should increase the same number. It doesn't feel like the HOLD is necessary except for item 10 since that is user-facing.
2. If we need a step before rollout plan item 2 to create a centralized method so that every step after can make use of it?

**Comment by JmillsExpensify** [comment]
> I'd say we can just close this since the last issue mentioned https://github.com/Expensify/App/issues/15352 should tackle this issue. If it hasn't then we reopened this issue later.

That's cool, it's just that we also need to pay out on this issue, and it's tied to a regression at the moment.

**Comment by PauloGasparSv** [comment]
Assigning the "Add a reportActionCount field to reports in Onyx" internal issue https://github.com/Expensify/App/issues/15353 to me here if that's ok!

I really like @chiragsalian suggestion to create a centralized function to fetch the report action count (like `getReportActionCount`) and I think we should add an issue for that.

**Comment by PauloGasparSv** [comment]
cc @roryabraham 

> Lastly, we'll update [ReportAPI::editComment](https://github.com/Expensify/Web-Expensify/blob/41873abd533de094c0e38387e6689b17c03e4548/lib/ReportAPI.php#L492-L623), so that, if we're deleting a comment we always update the report in Onyx to decrement the reportActionCount, regardless for whether it's the newest visible reportAction on the report, or if we shouldKeyReportActionsByID.

Hey, while testing https://github.com/Expensify/Web-Expensify/pull/36578 I realized that deleting a reportAction doesn't decrease `reportActionCount` in `GetChats.cpp`.

IMO that's ok but I want to know your opinions. If that's the case then I don't think we need to update the logic for deleting comments anymore since that counter doesn't decrease right?

**Comment by JmillsExpensify** [comment]
Still held for the linked tracking issue. Adding weekly back.

**Comment by PauloGasparSv** [comment]
No updates or blockers here, P.R. still in progress!

**Comment by PauloGasparSv** [comment]
Still no updates here.

**Comment by alex-mechler** [comment]
https://github.com/Expensify/App/issues/15952 is a dupe of this

**Comment by marcaaron** [comment]
Just out of curiosity, what do we still use `reportActionCount` for? Since we don't have `sequenceNumber` is there any value in it?

**Comment by marcaaron** [comment]
I thought we mainly just used it to calculate the unreads, but now that is something more like "last read compared to most recent created".

**Comment by trjExpensify** [comment]
@JmillsExpensify can we update the title to point to the issue it's held on for clarity? Pretty sure it's this one: https://github.com/Expensify/App/issues/15352.

**Comment by JmillsExpensify** [comment]
Agreed, I don't think this is one hold. This got wrapped up into the larger reportActionCount project.

**Comment by PauloGasparSv** [comment]
I'll have time to implement [this](https://github.com/Expensify/App/issues/14523#issuecomment-1512090552) between this and next week so I'll have an update here soon.

**Comment by michaelhaxhiu** [comment]
Since this is no longer on Hold, I'm going to flip it back to `Daily` per the standardized process (I'm just combing through for the weekly resync).

**Comment by PauloGasparSv** [comment]
Sure thing @michaelhaxhiu! No updates since [this](https://github.com/Expensify/App/issues/14523#issuecomment-1524054762) and this is still on my todo list.

**Comment by JmillsExpensify** [comment]
@roryabraham For confirmation, where did we land with implementing reportActionCount. I kind of remember a conversation where it might be paused?

**Comment by PauloGasparSv** [comment]
No updates here because I couldn't focus on this issue last week [as I planned](https://github.com/Expensify/App/issues/14523#issuecomment-1524054762). Now planning to work on this before EOW.

**Comment by PauloGasparSv** [comment]
No updates but we discussed this issue today [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1683757235958039).

**Comment by PauloGasparSv** [comment]
No updates

**Comment by PauloGasparSv** [comment]
Will start working on this tomorrow!

**Comment by PauloGasparSv** [comment]
I drafted the [P.R.](https://github.com/Expensify/App/pull/19321) and the changes seem to work! Still need to test more + add evidences and QA/Test steps before it is ready for reviews, will try do that today!

**Comment by PauloGasparSv** [comment]
> I drafted the https://github.com/Expensify/App/pull/19321 and the changes seem to work! Still need to test more + add evidences and QA/Test steps before it is ready for reviews, will try do that today!

Couldn't find time to get back to this but I'm finishing the P.R. before EOW.

**Comment by PauloGasparSv** [comment]
> Couldn't find time to get back to this but I'm finishing the P.R. before EOW.

Getting back to this Today!

**Comment by PauloGasparSv** [comment]
Still in progress, latest updates [here](https://github.com/Expensify/App/pull/19321#issuecomment-1569162554)

**Comment by PauloGasparSv** [comment]
Still making progress in the P.R. and still have some travis errors to fix, latest updates [here](https://expensify.slack.com/archives/C03U7DCU4/p1685732411516689?thread_ts=1685448724.492709&cid=C03U7DCU4)

**Comment by PauloGasparSv** [comment]
No new updates yet since [Friday](https://github.com/Expensify/App/issues/14523#issuecomment-1574414525), I'm getting back to this in a couple of hours to finish the P.R.

**Comment by PauloGasparSv** [comment]
Found some bugs and fixed them while testing. Right now there is [1 bug left](https://github.com/Expensify/App/pull/19321#issuecomment-1577710769) that I'm aware when you send a message and exit the chat really quickly before Onyx sends data back.

I'll try to fix that tomorrow.

**Comment by PauloGasparSv** [comment]
> Found some bugs and fixed them while testing. Right now there is [1 bug left](https://github.com/Expensify/App/pull/19321#issuecomment-1577710769) that I'm aware when you send a message and exit the chat really quickly before Onyx sends data back.
> 
> I'll try to fix that tomorrow.

Trying to reproduce this reliably to fix it!

**Comment by PauloGasparSv** [comment]
Was focused on a [Manual Requests issue](https://github.com/Expensify/Expensify/issues/278892) but will shift focus back to this tomorrow.

**Comment by PauloGasparSv** [comment]
P.R. still in progress and I'm fixing bugs, [more details there](https://github.com/Expensify/App/pull/19321#issuecomment-1590140672)!

I think I may need a couple of days to finish this but I'm aiming to make the P.R. ready for reviews tomorrow.

**Comment by PauloGasparSv** [comment]
Latest updates in the [P.R.](https://github.com/Expensify/App/pull/19321#issuecomment-1591977978), that bug I mentioned affected another P.R. that was discussed yesterday and we found a fix.

I'll apply the fix to this issue's P.R. and go back to testing asap

**Comment by PauloGasparSv** [comment]
Latest update [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1687197828706109?thread_ts=1687191765.942689&cid=C01GTK53T8Q)!

**Comment by PauloGasparSv** [comment]
No updates today but I'll shift focus back to this tomorrow.
(Focusing on https://github.com/Expensify/App/issues/19098 right now to get it done in time for SNH)

**Comment by PauloGasparSv** [comment]
> No updates today but I'll shift focus back to this tomorrow.
(Focusing on https://github.com/Expensify/App/issues/19098 right now to get it done in time for SNH)

Was still focused on finishing that issue today, will shift focus back to this tomorrow and try to get the P.R. in review.

**Comment by PauloGasparSv** [comment]
I made a small progress on the P.R. today after merging with main and adding the `currentReportId` fix mentioned [here](https://github.com/Expensify/App/issues/14523#issuecomment-1595268160) but since I focused most of my time these past 3 days on https://github.com/Expensify/App/issues/19098 and https://github.com/Expensify/Expensify/issues/290894 (for SNH and Manual Requests) I'll have to push putting the P.R. in review for tomorrow

**Comment by PauloGasparSv** [comment]
Pushing this for tomorrow once again.

I focused most of the day on this [NS problem](https://github.com/Expensify/Expensify/issues/287691) and now I'm focusing on the bug https://github.com/Expensify/App/issues/21609 where new users can't sign into newDot.

I'll go back to finish this issue's P.R. ASAP

**Comment by PauloGasparSv** [comment]
No updates today, will focus on this first thing in the morning tomorrow.

**Comment by PauloGasparSv** [comment]
Hey, I finish fixing the bugs I had reported last time and fixing the unit tests.

I re-tested everything in Web and only found a couple of issues, I described [both here](https://github.com/Expensify/App/pull/19321#issuecomment-1613756360) and I'm a little worried about the second problem so I'll describe it briefly:

When we create a new empty chat with someone, newDot generates a random `reportID for that chat. Because of that and the fact we are hiding empty report from LNH now, if someone goes offline they can spam a bunch of `OpenReport` calls to different `reportID`'s by re-creating the same chat over and over.

I'll investigate both issues further tomorrow.

**Comment by PauloGasparSv** [comment]
I prioritized [Manual Requests](https://github.com/Expensify/Expensify/issues/290894) and addressing some existing P.R. comments today over this so I didn't make any progress with the bugs I mentioned.

Will try to make progress tomorrow and will report here

**Comment by PauloGasparSv** [comment]
Still working on the bugs in that P.R., I managed to make some [progress today](https://github.com/Expensify/App/pull/19321#issuecomment-1624346160) but also introduced some new errors.

Will continue to prioritize this tomorrow

**Comment by PauloGasparSv** [comment]
Had no updates on Friday because I focused on https://github.com/Expensify/Expensify/issues/296893 and https://github.com/Expensify/Expensify/issues/297100, will try to focus on this again on Monday

**Comment by PauloGasparSv** [comment]
Focused on https://github.com/Expensify/Expensify/issues/296893 (discussed [here](https://expensify.slack.com/archives/C03TQ48KC/p1688680914512109)), P.R. reviews + addressing over this.

I don't think I'll have time to look into this today so doing it tomorrow.

**Comment by PauloGasparSv** [comment]
Forgot to give [this update](https://expensify.slack.com/archives/C056AV5V8UX/p1689206036958109?thread_ts=1688579181.045659&cid=C056AV5V8UX) here! 

Had to focus on Manual Requests + Deploy Blockers. I accumulated some newDot reviews from this week, once I'm done with those I'll work on this again.

**Comment by PauloGasparSv** [comment]
Not overdue!

**Comment by PauloGasparSv** [comment]
Hey team, I'm going OOO for 2 weeks (starting on July 17th, back on July 31st) so I won't be updating this issue until then!

cc @JmillsExpensify

**Comment by roryabraham** [comment]
Said another way: "hide empty chats from the LHN". Is that correct?

**Comment by JmillsExpensify** [comment]
Agreed. Floated this with an expert agency today. Should hear back in the coming days.

**Comment by pac-guerreiro** [comment]
Hey, I'm Pedro Guerreiro, an expert from Callstack, and I can take a look into this :)

**Comment by pac-guerreiro** [comment]
Sorry for the delay, I'll start working on this issue today 😄

**Comment by PauloGasparSv** [comment]
Hey I got back from OOO today!

@pac-guerreiro do you have any updates on this issue?

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv
> 
> Hey, I didn't make any advance with this issue since I had other issues in my plate sadly 😓
> 
> I did review your changes and noticed these issues:
> 
> 1. Going back/forward in browser history will show empty chats that were removed
> 2. Money request flow is broken
> 3. Empty chat report actions will remain persisted in local storage

Awesome, thks so much for the help here!

Btw, I also noticed we are now discussing this issue in this [slack thread](https://expensify.slack.com/archives/C03UK30EA1Z/p1690227330131039) so latest updates are there.

**Comment by PauloGasparSv** [comment]
As discussed [here](https://expensify.slack.com/archives/C03UK30EA1Z/p1691094158824919?thread_ts=1690227330.131039&cid=C03UK30EA1Z) I'll pick this back up from @pac-guerreiro

**Comment by PauloGasparSv** [comment]
Focused on other P.R.'s today so I'll pick this back up on Monday

**Comment by PauloGasparSv** [comment]
Hey, I've made a little bit of progress today, picked the P.R. back up, solved conflicts and started testing again to remember [the errors I was getting before going OOO](https://github.com/Expensify/App/issues/14523#issuecomment-1613761730).

I'm still focusing on the fact ReportId's are generated randomly in newDot and how to avoid duplicate reports and how to clear them from storage, I don't think that alone will be the solution).

I also did a quick investigation of [this issue](https://github.com/Expensify/App/issues/16883#issuecomment-1668683726) that was holding on [this issue's P.R.](https://github.com/Expensify/App/pull/19321).

**Comment by PauloGasparSv** [comment]
> I'm still focusing on the fact ReportId's are generated randomly in newDot and how to avoid duplicate reports and how to clear them from storage, I don't think that alone will be the solution).

Made more progress Today, turns out the duplicate report problem is not happening anymore! I fixed a couple of unrelated bugs and managed to re-test the chat/report creation correctly.

Now I'm investigating the Money Request flow if a user has an empty chat as that is causing the a `There is a previously existing chat between these users.` API error.

**Comment by PauloGasparSv** [comment]
Will continue working on this tomorrow

The money request error is coming from [here](https://github.com/Expensify/Auth/blob/d9029b3875504530e6a7be0fe2424f8cb3a0b443/auth/lib/Transaction.cpp#L1644) because, I just need to finish debugging the logic that makes the request in App to check if I have the correct `reportID` there to decide where to fix this in App or Auth.

**Comment by PauloGasparSv** [comment]
> Will continue working on this tomorrow
> 
> The money request error is coming from [here](https://github.com/Expensify/Auth/blob/d9029b3875504530e6a7be0fe2424f8cb3a0b443/auth/lib/Transaction.cpp#L1644) because, I just need to finish debugging the logic that makes the request in App to check if I have the correct `reportID` there to decide where to fix this in App or Auth.

I re-tested this and the issue is unrelated to my changes! The Money Request flow is also broken on `main`.

Now I'll test all possible chats here to see if there are any other bugs then put the P.R. for reviews

**Comment by PauloGasparSv** [comment]
P.R. is under review!

**Comment by JmillsExpensify** [comment]
Woo! It's live. Great work everyone getting this merged.

**Comment by JmillsExpensify** [comment]
@thesahindia Can you get the BZ checklist kicked off?

**Comment by JmillsExpensify** [comment]
Thanks, sorry I missed your last response. I definitely think we need to add a regression test for every new feature. As for the payment summary, does the above issue technically count as a regression? Otherwise, I believe the payment summary is simply:

 - Contributor: @bernhardoj $1,000
 - Contributor+: @thesahindia $1,000 (via NewDot).

**Comment by JmillsExpensify** [comment]
Ah thanks for your honesty! I'm not clear why you're assigned to this issue, though I'll update the payment summary in any case.

**Comment by JmillsExpensify** [comment]
Regression test has been created so I think we're all done here.

**Comment by JmillsExpensify** [comment]
$1,000 payment approved for @thesahindia based on BZ summary.
